reframe/frontend/dependency.py

```diff
@@ -85,25 +105,13 @@ def resolve_dep(target, from_map, fallback_map, *args):
         pname = c.partition.fullname
         ename = c.environ.name
         for dep in c.check.user_deps():
-            tname, how, subdeps = dep
-            if how == rfm.DEPEND_FULLY:
-                c.deps.extend(resolve_dep(c, cases_by_part,
-                                          default_cases_by_part, tname, pname))
-            elif how == rfm.DEPEND_BY_ENV:
-                c.deps.append(
-                    resolve_dep(c, cases_revmap, default_cases_revmap,
-                                tname, pname, ename)
-                )
-            elif how == rfm.DEPEND_EXACT:
-                for env, tenvs in subdeps.items():
-                    if env != ename:
-                        continue
-
-                    for te in tenvs:
-                        c.deps.append(
-                            resolve_dep(c, cases_revmap, default_cases_revmap,
-                                        tname, pname, te)
-                        )
+            tname, when = dep
+            for d in resolve_dep(c, all_cases, default_all_cases, tname):
```

Comment:
I think that the `resolve_dep` implementation can be simplified a lot. You only need to index per test name and this should return you all the test cases of that test (partition and environment). Then you simply add them as dependencies if the `when()` function returns `True`. Dependency errors should be raised only if the target test does not exist.